File: Data Structures & Algorithms/coin-change/submission-1.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<vector<int>> dp (n, vector<int>(amount + 1, 1e9));
        for(int i = 0; i < n; i++) dp[i][0] = 0;
        for(int j = 0; j <= amount; j++) {
            if (j % coins[0] == 0) dp[0][j] = (j/coins[0]);
        }
        
        for(int i = 1; i < n; i++) {
            for(int j = 0; j <= amount; j++) {
                int skip = dp[i-1][j];
                int take = 1e9;
                if(j >= coins[i]) {
                    take = 1 + dp[i][j - coins[i]];
                }
                dp[i][j] = min(skip, take);
            }
        }

        return dp[n-1][amount] if dp[n-1][amount] != 1e9 else -1;
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-1.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        def f(i, last):
            if i == n:
                return 0
            skip = f(i+1, last)
            take = 0
            if last == -1 or nums[i] > nums[last]:
                take = 1 + f(i+1, i)
            return max(skip, take)

        # return f(0, -1)
        dp = [1] * (n)
        for i in range(n-1, -1, -1):
            for j in range(i+1, n):
                if nums[i] < nums[j]:
                    dp[i] = max(dp[i], 1 + dp[j])
            
        return max(dp)

        